

From: Barak Avrahami <baraka@coregen.io>


---
 config.def.h |   55 ++++++++++++++++++++++++++++++++++++++++++++++++++++---
 dwm.c        |   46 ++++++++++++++++++++++++++++++++++++++++++----
 2 files changed, 94 insertions(+), 7 deletions(-)

diff --git a/config.def.h b/config.def.h
index 7f539ab..393478e 100644
--- a/config.def.h
+++ b/config.def.h
@@ -15,22 +15,71 @@ static const int systraypinningfailfirst = 1;   /* 1: if pinning fails, display
 static const int showsystray        = 1;     /* 0 means no systray */
 static const int showbar            = 1;     /* 0 means no bar */
 static const int topbar             = 1;     /* 0 means bottom bar */
+static const unsigned int colorfultitle  = 1;  /* 0 means title use SchemeTitle and SchemeTitleFloat */
+static const unsigned int colorfultag    = 1;  /* 0 means use SchemeSel for selected tag */
 static const char *fonts[]          = { "monospace:size=10" };
 static const char dmenufont[]       = "monospace:size=10";
 static const char col_gray1[]       = "#222222";
 static const char col_gray2[]       = "#444444";
 static const char col_gray3[]       = "#bbbbbb";
 static const char col_gray4[]       = "#eeeeee";
-static const char col_cyan[]        = "#005577";
+static const char col_bg[]          = "#1a1b26";
+static const char col_dark[]        = "#16161E";
+static const char col_fg[]          = "#a9b1d6";
+static const char col_black[]       = "#32344a";
+static const char col_red[]         = "#f7768e";
+static const char col_green[]       = "#9ece6a";
+static const char col_yellow[]      = "#e0af68";
+static const char col_blue[]        = "#7aa2f7";
+static const char col_magenta[]     = "#ad8ee6";
+static const char col_cyan[]        = "#449dab";
+static const char col_white[]       = "#787c99";
+static const char col_br_black[]    = "#444b6a";
+static const char col_br_red[]      = "#ff7a93";
+static const char col_br_green[]    = "#b9f27c";
+static const char col_br_yellow[]   = "#ff9e64";
+static const char col_br_blue[]     = "#7da6ff";
+static const char col_br_magenta[]  = "#bb9af7";
+static const char col_br_cyan[]     = "#0db9d7";
+static const char col_br_white[]    = "#acb0d0";
 static const char *colors[][3]      = {
 	/*               fg         bg         border   */
-	[SchemeNorm] = { col_gray3, col_gray1, col_gray2 },
-	[SchemeSel]  = { col_gray4, col_cyan,  col_cyan  },
+	[SchemeNorm]       = { col_gray3, col_gray1, col_gray2 },
+	[SchemeSel]        = { col_gray4, col_cyan,  col_cyan  },
+    [SchemeTag]        = { col_br_black,    col_bg,    col_black },
+    [SchemeTag1]       = { col_fg,          col_dark,  col_black },
+	[SchemeTag2]       = { col_red,         col_dark,  col_black },
+    [SchemeTag3]       = { col_br_yellow,   col_dark,  col_black },
+    [SchemeTag4]       = { col_br_blue,     col_dark,  col_black },
+    [SchemeTag5]       = { col_br_magenta,  col_dark,  col_black },
+	[SchemeTag6]       = { col_br_cyan,     col_dark,  col_black },
+	[SchemeTag7]       = { col_br_green,    col_dark,  col_black },
+	[SchemeTag8]       = { col_yellow,      col_dark,  col_black },
+	[SchemeTag9]       = { col_br_red,      col_dark,  col_black },
+	[SchemeLayout]     = { col_white,       col_bg,    col_black },
+	[SchemeTitle]      = { col_fg,          col_bg,    col_black },
+	[SchemeTitleFloat] = { col_br_blue,     col_bg,    col_black },
+	[SchemeTitle1]     = { col_fg,          col_bg,    col_black },
+	[SchemeTitle2]     = { col_red,         col_bg,    col_black },
+    [SchemeTitle3]     = { col_br_yellow,   col_bg,    col_black },
+    [SchemeTitle4]     = { col_br_blue,     col_bg,    col_black },
+    [SchemeTitle5]     = { col_br_magenta,  col_bg,    col_black },
+	[SchemeTitle6]     = { col_br_cyan,     col_bg,    col_black },
+	[SchemeTitle7]     = { col_br_green,    col_bg,    col_black },
+	[SchemeTitle8]     = { col_yellow,      col_bg,    col_black },
+	[SchemeTitle9]     = { col_br_red,      col_bg,    col_black },
 };
 
 /* tagging */
 static const char *tags[] = { "1", "2", "3", "4", "5", "6", "7", "8", "9" };
 
+static const int tagschemes[] = { SchemeTag1, SchemeTag2, SchemeTag3,
+                                  SchemeTag4, SchemeTag5, SchemeTag6,
+                                  SchemeTag7, SchemeTag8, SchemeTag9 };
+static const int titleschemes[] = { SchemeTitle1, SchemeTitle2, SchemeTitle3,
+                                    SchemeTitle4, SchemeTitle5, SchemeTitle6,
+                                    SchemeTitle7, SchemeTitle8, SchemeTitle9 };
+
 static const Rule rules[] = {
 	/* xprop(1):
 	 *	WM_CLASS(STRING) = instance, class
diff --git a/dwm.c b/dwm.c
index 49f816d..e5483c4 100644
--- a/dwm.c
+++ b/dwm.c
@@ -72,7 +72,16 @@
 
 /* enums */
 enum { CurNormal, CurResize, CurMove, CurLast }; /* cursor */
-enum { SchemeNorm, SchemeSel }; /* color schemes */
+
+enum { SchemeNorm, SchemeSel,
+       SchemeTag, SchemeTag1, SchemeTag2, SchemeTag3,
+       SchemeTag4, SchemeTag5, SchemeTag6, SchemeTag7,
+       SchemeTag8, SchemeTag9, SchemeLayout,
+       SchemeTitle, SchemeTitleFloat,
+       SchemeTitle1, SchemeTitle2, SchemeTitle3,
+       SchemeTitle4, SchemeTitle5, SchemeTitle6,
+       SchemeTitle7, SchemeTitle8, SchemeTitle9 }; /* color schemes */
+
 enum { NetSupported, NetWMName, NetWMState, NetWMCheck,
        NetSystemTray, NetSystemTrayOP, NetSystemTrayOrientation, NetSystemTrayOrientationHorz,
        NetWMFullscreen, NetActiveWindow, NetWMWindowType,
@@ -149,6 +158,8 @@ struct Monitor {
 	Monitor *next;
 	Window barwin;
 	const Layout *lt[2];
+    unsigned int colorfultitle;
+    unsigned int colorfultag;
 };
 
 typedef struct {
@@ -269,6 +280,8 @@ static int xerror(Display *dpy, XErrorEvent *ee);
 static int xerrordummy(Display *dpy, XErrorEvent *ee);
 static int xerrorstart(Display *dpy, XErrorEvent *ee);
 static void zoom(const Arg *arg);
+static void togglecolorfultitle();
+static void togglecolorfultag();
 
 /* variables */
 static Systray *systray = NULL;
@@ -732,6 +745,8 @@ createmon(void)
 	m->nmaster = nmaster;
 	m->showbar = showbar;
 	m->topbar = topbar;
+    m->colorfultag = colorfultag ? colorfultag : 0;
+    m->colorfultitle = colorfultitle ? colorfultitle : 0;
 	m->gappih = gappih;
 	m->gappiv = gappiv;
 	m->gappoh = gappoh;
@@ -935,7 +950,7 @@ drawbar(Monitor *m)
 	x = 0;
 	for (i = 0; i < LENGTH(tags); i++) {
 		w = TEXTW(tags[i]);
-		drw_setscheme(drw, scheme[m->tagset[m->seltags] & 1 << i ? SchemeSel : SchemeNorm]);
+        drw_setscheme(drw, scheme[m->tagset[m->seltags] & 1 << i ? (m->colorfultag ? tagschemes[i] : SchemeSel) : SchemeTag]);
 		drw_text(drw, x, 0, w, bh, lrpad / 2, tags[i], urg & 1 << i);
 		if (occ & 1 << i)
 			drw_rect(drw, x + boxs, boxs, boxw, boxw,
@@ -944,12 +959,21 @@ drawbar(Monitor *m)
 		x += w;
 	}
 	w = blw = TEXTW(m->ltsymbol);
-	drw_setscheme(drw, scheme[SchemeNorm]);
+    drw_setscheme(drw, scheme[SchemeLayout]);
 	x = drw_text(drw, x, 0, w, bh, lrpad / 2, m->ltsymbol, 0);
 
 	if ((w = m->ww - tw - stw - x) > bh) {
 		if (m->sel) {
-			drw_setscheme(drw, scheme[m == selmon ? SchemeSel : SchemeNorm]);
+            if (selmon->colorfultitle) {
+                for (i = 0; i < LENGTH(tags); i++)
+                    if (selmon->sel->tags & 1 << i)
+                        drw_setscheme(drw, scheme[titleschemes[i]]);
+                    } else {
+                        int s = (m == selmon) && m->sel->isfloating
+                                        ? SchemeTitleFloat
+                                        : SchemeTitle;
+                        drw_setscheme(drw, scheme[s]);
+                    }
 			drw_text(drw, x, 0, w, bh, lrpad / 2, m->sel->name, 0);
 			if (m->sel->isfloating)
 				drw_rect(drw, x + boxs, boxs, boxw, boxw, m->sel->isfixed, 0);
@@ -2018,6 +2042,20 @@ togglefloating(const Arg *arg)
 	arrange(selmon);
 }
 
+void
+togglecolorfultitle()
+{
+        selmon->colorfultitle = !selmon->colorfultitle;
+        drawbar(selmon);
+}
+
+void
+togglecolorfultag()
+{
+        selmon->colorfultag = !selmon->colorfultag;
+        drawbar(selmon);
+}
+
 void
 toggletag(const Arg *arg)
 {
